File: h04/handson4_3.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Sep 18 13:53:32 2017

@author: luciana
"""

import numpy as np
import time
## Parâmetros do sinal 
fc=0.04                              # Frequência do seno
Fs=1                                 # Frequência de amostragem
Ts = 1/Fs                            # Tempo entre amostras
A = 10                               # Amplitude do sinal
nC = 2000                            # Número de períodos da onda   
t=np.arange(0,nC/fc,Ts)              # Vetor tempo
x=A*np.cos(2*np.pi*fc*t)             # Gera o sinal x(n)
N = len(x)                           # Número de amostras do sinal

sT1 = time.clock() 
Px_tempo =(np.linalg.norm(x)**2)/N   # Cálculo da potência no tempo
print('Potência via amostras no tempo = ',Px_tempo,time.clock()-sT1,' segundos para calcular.')

sT2 =time.clock()
Nfft = 1000                          # Número de pontos da FFT
Xfft=np.fft.fft(x,Nfft)              # Encontra a FFT
Xfft = Xfft/Nfft                     # Encontra a FFT 
Px_fft = np.sum(Xfft*np.conj(Xfft))  # Cálculo da potência na frêquencia
print('Potência via FFT (Parseval) = ',abs(Px_fft),time.clock()-sT2,' segundos para calcular.' )

sT3 =time.clock()
Rxx=np.correlate(x,x,'full')/len(x)  # Estima a autocorrelaçao de x(n)
Px_Rxx = Rxx[N-1]                    # Cálculo da potência duas bandas via autocorrelação
print('Potência via autocorrelação = ',Px_Rxx,time.clock()-sT3, ' segundos para calcular.' )

sT4 =time.clock()
Pvar = np.var(x)                     # Cálculo da potência duas bandas via variancia
print('Potência via variância = ',Pvar,time.clock()-sT4,' segundos para calcular.' )

File: h04/handson4_3.m
clc; close all;clear all;
%% Parâmetros do sinal 
fc=0.04;                              % Frequência do seno
Fs=10;                                % Frequência de amostragem
Ts = 1/Fs;                            % Tempo entre amostras
A = 10;                               % Amplitude do sinal
%nC = 2000;                            % Número de períodos da onda   
nC = 20000;                            % Número de períodos da onda   
t=0:Ts:nC/fc-Ts;                      % Vetor tempo
x=A*cos(2*pi*fc*t);                   % Gera o sinal x(n)
N = length(x);                        % Número de amostras do sinal

sT1 = tic; 
Px_tempo =(norm(x)^2)/N; % Cálculo da potência no tempo
disp(['Potência via amostras no tempo = ' num2str(Px_tempo) ', ' num2str(toc(sT1)) ' segundos para calcular.' ]);

sT2 = tic;
Nfft = 1000;                          % Número de pontos da FFT
Xfft=fft(x,Nfft);                     % Encontra a FFT
Xfft = Xfft/Nfft;                     % Encontra a FFT 
Px_fft = sum(Xfft.*conj(Xfft));       % Cálculo da potência na frêquencia
disp(['Potência via FFT (Parseval) = ' num2str(Px_fft) ', ' num2str(toc(sT2)) ' segundos para calcular.' ]);

sT3 = tic;
Rxx=xcorr(x,'biased');                % Estima a autocorrelaçao de x(n)
Px_Rxx = Rxx(N);                      % Cálculo da potência duas bandas via autocorrelação
disp(['Potência via autocorrelação = ' num2str(Px_Rxx) ', ' num2str(toc(sT3)) ' segundos para calcular.' ]);

sT4 = tic;
Px_bPower = bandpower(x);             % Cálculo da potência via bandpower
disp(['Potência via função bandpower = ' num2str(Px_bPower) ', ' num2str(toc(sT4)) ' segundos para calcular.' ]);

sT5 = tic;
Pvar = var(x);                        % Cálculo da potência duas bandas via variancia
disp(['Potência via variância = ' num2str(Pvar) ', ' num2str(toc(sT5)) ' segundos para calcular.' ]);